src/vs/workbench/browser/parts/views/customView.ts

```diff
@@ -780,33 +780,27 @@ class TreeRenderer extends Disposable implements ITreeRenderer<ITreeItem, FuzzyS
 		// reset
 		templateData.actionBar.clear();
 
-		if (resource || node.themeIcon) {
+		if (resource || this.isFileKindThemeIcon(node.themeIcon)) {
 			const fileDecorations = this.configurationService.getValue<{ colors: boolean, badges: boolean }>('explorer.decorations');
 			templateData.resourceLabel.setResource({ name: label, description, resource: resource ? resource : URI.parse('missing:_icon_resource') }, { fileKind: this.getFileKind(node), title, hideIcon: !!iconUrl, fileDecorations, extraClasses: ['custom-view-tree-node-item-resourceLabel'], matches: matches ? matches : createMatches(element.filterData) });
 		} else {
 			templateData.resourceLabel.setResource({ name: label, description }, { title, hideIcon: true, extraClasses: ['custom-view-tree-node-item-resourceLabel'], matches: matches ? matches : createMatches(element.filterData) });
 		}
 
 		templateData.icon.title = title ? title : '';
-		DOM.toggleClass(templateData.icon, 'custom-view-tree-node-item-icon', !!iconUrl);
 
-		let codicon: string | undefined;
-		if (iconUrl?.scheme === 'vscode-icon' && iconUrl?.authority === 'codicon') {
-			codicon = `codicon-${escape(iconUrl.path.substr(1))}`;
-		}
+		if (iconUrl) {
+			templateData.icon.className = 'custom-view-tree-node-item-icon';
+			templateData.icon.style.backgroundImage = DOM.asCSSUrl(iconUrl);
 
-		DOM.toggleClass(templateData.icon, 'codicon', !!codicon);
-		templateData.icon.classList.forEach((cls, i, list) => {
-			if (cls !== codicon && cls.indexOf('codicon-') === 0) {
-				list.remove(cls);
+		} else {
+			let iconClass: string | undefined;
+			if (node.themeIcon && !this.isFileKindThemeIcon(node.themeIcon)) {
+				iconClass = ThemeIcon.asClassName(node.themeIcon);
 			}
-		});
-		if (codicon) {
-			DOM.addClass(templateData.icon, codicon);
+			templateData.icon.className = iconClass ? `custom-view-tree-node-item-icon ${iconClass}` : '';
```

Comment:
@alexr00 fyi and for another pair of eyes. This adds support for ThemeIcons that are different than the file & folder constant, e.g `new ThemeIcon('zap')`

Reply:
I don't think this will correctly handle the codicon url that @eamodio added. Instead of checking for `iconUrl` you might need to check for the scheme and authority as it currently does. 
It looks like the icon class will now always include `custom-view-tree-node-item-icon` if there's any icon, whereas before it would only have that if there was an `iconUrl`. Might have just been an oversight in 1ba624b7b455d46c311b6a5ba85620044f9bb64b though.

Reply:
The plan is that this replaces the codicon-uri so that uris are used for resources on disk/web and that ThemeIcon is used for codicon

Reply:
Love it. It is much simpler.